File: daily/cpp/232_implement_queue_using_stacks.cpp
#include <bits/stdc++.h>
using namespace std;

class MyQueue {
public:
    MyQueue() {
        while (!s1.empty()) s1.pop();
        while (!s2.empty()) s2.pop();
    }

    void push(int x) {
        s1.push(x);
    }

    int pop() {
        if (s2.empty()) {
            while (!s1.empty()) {
                int x = s1.top();
                s1.pop();
                s2.push(x);
            }
        }
        int x = s2.top();
        s2.pop();
        return x;
    }

    int peek() {
        if (s2.empty()) {
            while (!s1.empty()) {
                int x = s1.top();
                s1.pop();
                s2.push(x);
            }
        }
        return s2.top();
    }

    bool empty() {
        return s1.empty() && s2.empty();
    }

private:
    stack<int> s1, s2;
};

File: daily/rust/232_implement_queue_using_stacks.rs
struct MyQueue {
    stack: Vec<i32>,
    temp: Vec<i32>,
}

impl MyQueue {
    fn new() -> Self {
        MyQueue {
            stack: Vec::new(),
            temp: Vec::new(),
        }
    }

    fn push(&mut self, x: i32) {
        while let Some(v) = self.stack.pop() {
            self.temp.push(v);
        }
        self.stack.push(x);
        while let Some(v) = self.temp.pop() {
            self.stack.push(v);
        }
    }

    fn pop(&mut self) -> i32 {
        self.stack.pop().unwrap()
    }

    fn peek(&self) -> i32 {
        *self.stack.last().unwrap()
    }

    fn empty(&self) -> bool {
        self.stack.is_empty()
    }
}